feat: kernel hub introduction draft (#2777)
huggingface/blog: hello-hf-kernels.md

---
title: "Learn the Hugging Face Kernel Hub in 5 Minutes"
thumbnail: /blog/assets/hello-hf-kernels/kernel-hub-five-mins-short.png
authors:
- user: drbh

> date: 2025-03-28 

Author: thanks! updated in the latest commits

---

# 🏎️ Learn the Hugging Face Kernel Hub in 5 Minutes

> **Unlock performance boosts for your models with pre-optimized compute kernels, easily loaded from the Hub.**

Author: thanks! updated in the latest commits

> Today, we'll explore an exciting development from Hugging Face: the **Kernel Hub**! As ML practitioners, we know that maximizing performance often involves diving deep into optimized code, custom CUDA kernels, or complex build systems. The Kernel Hub aims to simplify this dramatically.

Author: oh this is better, updated in latest commit

We'll cover the following topics:

1.  **What is the Kernel Hub?** - Understanding the core concept.
2.  **How to use the Kernel Hub** - A quick code example.
3.  **Adding a Kernel to a Simple Model** - A practical integration using RMSNorm.
4.  **Reviewing Performance Impact** - Benchmarking the RMSNorm difference.

> And we'll introduce these concepts quickly – the core idea can be grasped in about 5 minutes (though experimenting and benchmarking might take a bit longer!).

Reviewer: ```suggestion
We'll introduce these concepts quickly – the core idea can be grasped in about 5 minutes (though experimenting and benchmarking might take a bit longer!).
```

Author: also better! thanks! updated

## 1. What is the Kernel Hub?

> The [Kernel Hub](https://huggingface.co/kernels) (👈 Check it out!) allows Python libraries and applications to **load optimized compute kernels directly from the Hugging Face Hub**. Think of it like the Model Hub, but for low-level, high-performance code snippets (kernels) that accelerate specific operations, often on GPUs. Examples include optimized attention mechanisms (like FlashAttention), activation functions, and normalization layers (like LayerNorm or RMSNorm).

Reviewer 2: I think it would be better to mention some challenging kernels here. I think activation and normalization kernels are usually pretty good in frameworks. Maybe, attention mechanisms, quantizers, and Mixture of Expert layers?

> The [Kernel Hub](https://huggingface.co/kernels) (👈 Check it out!) allows Python libraries and applications to **load optimized compute kernels directly from the Hugging Face Hub**. Think of it like the Model Hub, but for low-level, high-performance code snippets (kernels) that accelerate specific operations, often on GPUs. Examples include optimized attention mechanisms (like FlashAttention), activation functions, and normalization layers (like LayerNorm or RMSNorm).

Author: yep, good catch. updated in latest

Instead of manually managing complex dependencies, wrestling with compilation flags, or building libraries like Triton or CUTLASS from source, you can use the `kernels` library to instantly fetch and run pre-compiled, optimized kernels.

### Benefits of the Kernel Hub:

> * **Instant Access to Optimized Kernels**: Load and run kernels optimized for various hardware (like NVIDIA GPUs) without local compilation hassles.

Reviewer 3: ```suggestion
* **Instant Access to Optimized Kernels**: Load and run kernels optimized for various hardware starting with NVIDIA and AMD GPUs, without local compilation hassles.
```

* **Share and Reuse**: Discover, share, and reuse kernels across different projects and the community.
* **Easy Updates**: Stay up-to-date with the latest kernel improvements simply by pulling the latest version from the Hub.
* **Accelerate Development**: Focus on your model architecture and logic, not on the intricacies of kernel compilation and deployment.
* **Improve Performance**: Leverage kernels optimized by experts to potentially speed up training and inference.

> * **Simplify Deployment**: Reduce the complexity of your deployment environment by fetching kernels on demand.

Reviewer: Maybe another point could be about sharing your kernels with the community, for those inclined to develop new ones.

Author: agreed, added this bullet

```
* **Develop and Share Your Own Kernels**: If you create optimized kernels, you can easily share them on the Hub for others to use. This encourages collaboration and knowledge sharing within the community.
```

> As many machine learning developers know, managing dependencies and building low-level code from source can be a time-consuming and error-prone process. The Kernel Hub aims to simplify this by providing a centralized repository of optimized compute kernels that can be easily loaded and run.

Spend more time building great models and less time fighting build systems!

## 2. How to Use the Kernel Hub (Basic Example)

> Using the Kernel Hub is designed to be straightforward. The `kernels` library provides the main interface. Here's a quick example loading an optimized GELU activation function kernel (we'll use a different kernel for the main example later).

Reviewer: ```suggestion
Using the Kernel Hub is designed to be straightforward. The `kernels` library provides the main interface. Here's a quick example that loads an optimized GELU activation function kernel. (Later on, we'll see another example about how to integrate a kernel in our model).
```

Author: thanks this reads better, updated in latest

~~~python
# /// script
# dependencies = [
#  "numpy",
#  "torch",
#  "kernels",
# ]
# ///

import torch
from kernels import get_kernel

# Ensure you have a CUDA-enabled device
if not torch.cuda.is_available():

>     raise RuntimeError("This example requires a CUDA-enabled GPU")

Reviewer 2: Let me upload the `activation` kernel for ROCm as well. I think the example is stronger if we can show something that works with both CUDA and ROCm.

Reviewer 2: https://huggingface.co/kernels-community/activation/tree/main/build/torch26-cxx11-rocm62-x86_64-linux

Built, running validation tests now...

Reviewer 2: All tests pass.

Author: wooo amazing, thank you!

DEVICE = "cuda"

# Make reproducible
torch.manual_seed(42)

# Download optimized activation kernels from the Hub
# This fetches the kernel code if not already cached

> activation_kernels = get_kernel("kernels-community/activation")

Reviewer: Super cool! Would something like [this](https://huggingface.co/kernels-community/mel_spectrogram/tree/main/build) (different kernel) be automatically resolved? Do we want to talk (in a later section) about what happens if there's no match?

# Create a random tensor on the GPU
x = torch.randn((4, 4), dtype=torch.float16, device=DEVICE)

# Prepare an output tensor
y = torch.empty_like(x)

# Run the specific kernel function (e.g., fast GELU)
# The `activation_kernels` object holds multiple functions
activation_kernels.gelu_fast(y, x)

# Check the output against expected values
expected = torch.tensor(
    [
        [0.1100, 2.1309, -0.0700, 0.6802],
        [-0.0500, 0.4800, -0.1700, -0.1700],
        [0.3701, -0.1300, -0.0800, -0.1200],
        [-0.0400, 0.1200, -0.1500, 1.7998],
    ],
    dtype=torch.float16,
    device=DEVICE,

> )

Reviewer: Perhaps an alternative could be to retrieve the reference results from PyTorch's gelu?

Author: yea agreed that is a better example, updated in latest commit

3.  **Prepare Tensors**: We create input (`x`) and output (`y`) tensors on the GPU.
4.  **`activation_kernels.gelu_fast(y, x)`**: We call the specific optimized function (`gelu_fast`) provided by the loaded kernel module.
5.  **Verification**: We check the output.

This simple example shows how easily you can fetch and execute highly optimized code. Now let's look at a more practical integration using RMS Normalization.

## 3. Add a Kernel to a Simple Model

Let's integrate an optimized **RMS Normalization** kernel into a basic model. We'll use the `LlamaRMSNorm` implementation provided in the `kernels-community/triton-layer-norm` repository (note: this repo contains various normalization kernels) and compare it against a baseline PyTorch implementation of RMSNorm.

First, define a simple RMSNorm module in PyTorch and a baseline model using it:

~~~python
# /// script
# dependencies = [
#  "numpy",
#  "torch",
#  "kernels",
# ]
# ///
import torch
import torch.nn as nn

DEVICE = "cuda"
if not torch.cuda.is_available():

>     raise RuntimeError("This example requires a CUDA-enabled GPU")

Reviewer 2: I think the Triton kernel should also work with ROCm? Worth trying.

Author: awesome, thanks for building/testing! removed `torch.cuda..` in the latest commit

baseline_model = (
    BaselineModel(input_size, hidden_size, output_size, eps=eps_val)
    .to(DEVICE)
    .to(DTYPE)
)
dummy_input = torch.randn(32, input_size, device=DEVICE, dtype=DTYPE)  # Batch of 32
output = baseline_model(dummy_input)
print("Baseline RMSNorm model output shape:", output.shape)
~~~

Now, let's create a version using the `LlamaRMSNorm` kernel loaded via `kernels`.

~~~python
# /// script
# dependencies = [
#  "numpy",
#  "torch",
#  "kernels",
# ]
# ///
import torch
import torch.nn as nn
from kernels import get_kernel

# reuse the model from the previous snippet or copy the class
# definition here to run this script independently
from snippet2 import BaselineModel

DEVICE = "cuda"
if not torch.cuda.is_available():
    raise RuntimeError("This example requires a CUDA-enabled GPU")
DTYPE = torch.float16  # Use float16 for better kernel performance potential

layer_norm_kernel_module = get_kernel("kernels-community/triton-layer-norm")

class KernelRMSNorm(layer_norm_kernel_module.layers.LlamaRMSNorm):
    def __init__(self, hidden_size, variance_epsilon=1e-5):
        super().__init__()
        self.weight = nn.Parameter(torch.ones(hidden_size))

>         self.variance_epsilon = variance_epsilon

Reviewer 2: This is kinda the opposite of the pattern that we want people to use :). They should use `@use_kernel_forward_from_hub` to annotate the Torch class and then register `LlamaRMSNorm` using a mapping. See: https://github.com/huggingface/kernels/blob/main/docs/layers.md

Using `@use_kernel_forward_from_hub` enables people to make layers that are (dynamically) extensible with kernels, people can replace kernels, etc.

Author: ah yea great point! I've updated the code to prefer adding `@use_kernel_forward_from_hub("LlamaRMSNorm")` to the `RMSNorm` defined in the reference example (and added some descriptive comments).

class KernelModel(nn.Module):
    def __init__(
        self,
        input_size,
        hidden_size,
        output_size,
        device="cuda",
        dtype=torch.float16,
        eps=1e-5,
    ):
        super().__init__()
        self.linear1 = nn.Linear(input_size, hidden_size)

>         self.norm = KernelRMSNorm(hidden_size, variance_epsilon=eps)

Reviewer 2: With `@use_kernel_forward_from_hub`, you don't need this. The model doesn't need any change to use kernels, the model writer or the user can map kernels externally.

        input_size, hidden_size, output_size, device=DEVICE, dtype=DTYPE, eps=eps_val
    )
    .to(DEVICE)
    .to(DTYPE)
)

baseline_model = (
    BaselineModel(input_size, hidden_size, output_size, eps=eps_val)
    .to(DEVICE)
    .to(DTYPE)
)

dummy_input = torch.randn(32, input_size, device=DEVICE, dtype=DTYPE)  # Batch of 32

output = baseline_model(dummy_input)
output_kernel = kernel_model(dummy_input)
print("Kernel RMSNorm model output shape:", output_kernel.shape)

# Verify outputs are close (RMSNorm implementations should be numerically close)
try:
    torch.testing.assert_close(output, output_kernel, rtol=1e-2, atol=1e-2)
    print("\nBaseline and Kernel RMSNorm model outputs match!")
except AssertionError as e:
    print("\nBaseline and Kernel RMSNorm model outputs differ slightly:")
    print(e)
except NameError:
    print("\nSkipping output comparison as kernel model output was not generated.")

~~~

**Important Notes on the `KernelModel`:**
* **Kernel Inheritance:** The `KernelRMSNorm` class inherits from `layer_norm_kernel_module.layers.LlamaRMSNorm`, which is the RMSNorm implementation in the kernel. This allows us to use the optimized kernel directly.

> * **Accessing the Function:** The exact way to access the RMSNorm function (`layer_norm_kernel_module.layers.LlamaRMSNorm.forward`, `layer_norm_kernel_module.rms_norm_forward`, or something else) **depends entirely on how the kernel creator structured the repository on the Hub.** You may need to inspect the loaded `layer_norm_kernel_module` object (e.g., using `dir()`) or check the kernel's documentation on the Hub to find the correct function/method and its signature. I've used `rms_norm_forward` as a plausible placeholder and added error handling.

Reviewer: Would be nice if we can point to some kernel documentation (in the kernel's model card in the Hub) by the time this is published :) This could encourage others to adopt some common structure for kernel description / docs.

Author: agreed! currently there is a effort to generate some useful docs started here https://github.com/huggingface/kernel-builder/pull/89 however this is still a work in progress and should be updated before publishing

## TODO
- [ ] improve docs across all existing examples (probably autogen)

* **Parameters:** We now only define `rms_norm_weight` (no bias), consistent with RMSNorm.

## 4. Review Performance Impact

Does using the optimized Triton RMSNorm kernel provide a speedup compared to the basic PyTorch version? Let's benchmark the forward pass again.

~~~python
# /// script
# dependencies = [
#  "numpy",
#  "torch",
#  "kernels",
# ]
# ///
import torch

# reuse the models from the previous snippets or copy the class
# definitions here to run this script independently
from snippet2 import BaselineModel

> from snippet3 import KernelModel

Reviewer: We should introduce the script name before each snippet, I think.

Author: good point, updated to have meaningful names and use them in the scripts in latest

* Quality of the specific kernel implementation.
* Potential overhead for small inputs.

Actual results will depend on your hardware and the specific kernel implementation. Here's an example of what you might see (on a L4 GPU):
```txt
Batch Size   | Baseline Time (ms) | Kernel Time (ms)   | Speedup
--------------------------------------------------------------------------
256          | 0.2122             | 0.2911             | 1.37x slower
512          | 0.4748             | 0.3312             | 1.43x
1024         | 0.8946             | 0.6864             | 1.30x
2048         | 2.0289             | 1.3889             | 1.46x
4096         | 4.4318             | 2.2467             | 1.97x
8192         | 9.2438             | 4.8497             | 1.91x
16384        | 18.6992            | 9.8805             | 1.89x
32768        | 37.079             | 19.9461            | 1.86x
65536        | 73.588             | 39.593             | 1.86x
```

## Get Started and Next Steps!

You've seen how easy it is to fetch and use optimized kernels with the Hugging Face Kernel Hub. Ready to try it yourself?

1.  **Install the library:**
    ~~~bash
    pip install kernels torch numpy
    ~~~

>     Ensure you have a compatible PyTorch version and CUDA installed if using GPU kernels.

Author: good catch, i've updated the phrasing to avoid "CUDA" in the latest commit

### Benefits of the Kernel Hub:

* **Instant Access to Optimized Kernels**: Load and run kernels optimized for various hardware starting with NVIDIA and AMD GPUs, without local compilation hassles.
* **Share and Reuse**: Discover, share, and reuse kernels across different projects and the community.
* **Easy Updates**: Stay up-to-date with the latest kernel improvements simply by pulling the latest version from the Hub.
* **Accelerate Development**: Focus on your model architecture and logic, not on the intricacies of kernel compilation and deployment.
* **Improve Performance**: Leverage kernels optimized by experts to potentially speed up training and inference.
* **Simplify Deployment**: Reduce the complexity of your deployment environment by fetching kernels on demand.
* **Develop and Share Your Own Kernels**: If you create optimized kernels, you can easily share them on the Hub for others to use. This encourages collaboration and knowledge sharing within the community.

> > As many machine learning developers know, managing dependencies and building low-level code from source can be a time-consuming and error-prone process. The Kernel Hub aims to simplify this by providing a centralized repository of optimized compute kernels that can be easily loaded and run.

Reviewer 4: this is a quote format btw!

Spend more time building great models and less time fighting build systems!

## 2. How to Use the Kernel Hub (Basic Example)

Using the Kernel Hub is designed to be straightforward. The `kernels` library provides the main interface. Here's a quick example that loads an optimized GELU activation function kernel. (Later on, we'll see another example about how to integrate a kernel in our model).

> File: `activation_validation_example.py`

Reviewer 4: I'd add a link to the file so it's easy for people to directly check

Author: great point, ive made all of the files gists and added links to them! thanks

except AssertionError as e:
    print("\nBaseline and Kernel RMSNorm model outputs differ slightly:")
    print(e)
except NameError:
    print("\nSkipping output comparison as kernel model output was not generated.")

~~~

**Important Notes on the `KernelModel`:**

* **Kernel Inheritance:** The `KernelRMSNorm` class inherits from `layer_norm_kernel_module.layers.LlamaRMSNorm`, which is the RMSNorm implementation in the kernel. This allows us to use the optimized kernel directly.
* **Accessing the Function:** The exact way to access the RMSNorm function (`layer_norm_kernel_module.layers.LlamaRMSNorm.forward`, `layer_norm_kernel_module.rms_norm_forward`, or something else) **depends entirely on how the kernel creator structured the repository on the Hub.** You may need to inspect the loaded `layer_norm_kernel_module` object (e.g., using `dir()`) or check the kernel's documentation on the Hub to find the correct function/method and its signature. I've used `rms_norm_forward` as a plausible placeholder and added error handling.
* **Parameters:** We now only define `rms_norm_weight` (no bias), consistent with RMSNorm.

## 4. Review Performance Impact

> Does using the optimized Triton RMSNorm kernel provide a speedup compared to the basic PyTorch version? Let's benchmark the forward pass again.

Reviewer 4: ```suggestion
Does optimized Triton RMSNorm kernel speeds up compared to the kernel in basic PyTorch? Let's benchmark the forward pass again.
```
sentence felt a bit hard to read, rephrased (if you feel like it)

Author: good catch thanks for the suggestion, I ended up rewriting that part to

---

## 4. Benchmarking the Performance Impact

How much faster is the optimized Triton RMSNorm kernel compared to the standard PyTorch version? Let’s benchmark the forward pass to find out.

File: [`rmsnorm_benchmark.py`](https://gist.github.com/drbh/c754a4ba52bcc46190ae4a45516fb190)

...

    kernel_time = -1  # Sentinel value

    kernel_time = benchmark_model(kernel_model_bench, bench_input)

    baseline_time = round(baseline_time, 4)
    kernel_time = round(kernel_time, 4)
    speedup = round(baseline_time / kernel_time, 2) if kernel_time > 0 else "N/A"
    if kernel_time < baseline_time:
        speedup = f"{speedup:.2f}x"
    elif kernel_time == baseline_time:
        speedup = "1.00x (identical)"
    else:
        speedup = f"{kernel_time / baseline_time:.2f}x slower"
    print(f"{batch_size:<12} | {baseline_time:<18} | {kernel_time:<18} | {speedup}")

~~~

**Expected Outcome:**
Similar to LayerNorm, optimized RMSNorm kernels (especially those using Triton) implemented for specific hardware can offer significant speedups over basic PyTorch implementations, particularly for memory-bound operations on suitable hardware (e.g., NVIDIA Ampere/Hopper GPUs) and data types (`float16`/`bfloat16`).

**Important Caveats (Remain Applicable):**
* Microbenchmark limitations.
* Dependence on Hardware, Input Size, Dtype.
* Quality of the specific kernel implementation.
* Potential overhead for small inputs.

> Actual results will depend on your hardware and the specific kernel implementation. Here's an example of what you might see (on a L4 GPU):

Reviewer 4: <img width="823" alt="Screenshot 2025-05-02 at 11 39 27" src="https://github.com/user-attachments/assets/dd78caeb-b0dd-472f-b362-f2cbbea219ca" />
heads up, it looks like this. you can render them here hf.co/new-blog when in doubt

Author: thank you! fixed to be a table in the latest changes 🙏

> ```txt

Reviewer 4: fixes above format
```suggestion

```txt
```

---
title: "Learn the Hugging Face Kernel Hub in 5 Minutes"
thumbnail: /blog/assets/hello-hf-kernels/kernel-hub-five-mins-short.png
authors:
- user: drbh
- user: danieldk
- user: narsil
- user: pcuenca
- user: pagezyhf
- user: merve
date: 2025-03-28 
---

> # 🏎️ Learn the Hugging Face Kernel Hub in 5 Minutes

Author: ah yea that a good call out, I've updated the header to "# 🏎️ Enhance Your Models in 5 Minutes with the Hugging Face Kernel Hub" which is a bit more descriptive.. but happy to change to anything better!

**Boost your model performance with pre-optimized kernels, easily loaded from the Hub.**

Today, we'll explore an exciting development from Hugging Face: the **Kernel Hub**! As ML practitioners, we know that maximizing performance often involves diving deep into optimized code, custom CUDA kernels, or complex build systems. The Kernel Hub simplifies this process dramatically!

> 

Author: great point, Ive added a very short snippet of a copy/paste example which should be a better hook for devs

We'll cover the following topics:

1.  **What is the Kernel Hub?** - Understanding the core concept.
2.  **How to use the Kernel Hub** - A quick code example.
3.  **Adding a Kernel to a Simple Model** - A practical integration using RMSNorm.
4.  **Reviewing Performance Impact** - Benchmarking the RMSNorm difference.

> We'll introduce these concepts quickly – the core idea can be grasped in about 5 minutes (though experimenting and benchmarking might take a bit longer!).

Author: very good point! just added ("## 5. Real World Use Cases") and linked to where they are used and totally agree this strengthens the article 🙏 thank you!

## 1. What is the Kernel Hub?

The [Kernel Hub](https://huggingface.co/kernels-community) (👈 Check it out!) allows Python libraries and applications to **load optimized compute kernels directly from the Hugging Face Hub**. Think of it like the Model Hub, but for low-level, high-performance code snippets (kernels) that accelerate specific operations, often on GPUs. 

> Examples include advanced attention mechanisms (like [FlashAttention](https://huggingface.co/kernels-community/flash-attn) for dramatic speedups and memory savings). Custom [quantization kernels](https://huggingface.co/kernels-community/quantization) (enabling efficient computation with lower-precision data types like INT8 or INT4). Specialized kernels required for complex architectures like [Mixture of Experts (MoE) layers](https://huggingface.co/kernels-community/moe), which involve intricate routing and computation patterns. As well as [activation functions](https://huggingface.co/kernels-community/activation), and [normalization layers (like LayerNorm or RMSNorm)](https://huggingface.co/kernels-community/triton-layer-norm).

Author: ah yea that should be resolved soon as the kernel-builder jobs integration is still experimental (to auto generate builds). It should be stable soon, however if this takes awhile I'll update the readme. 

PS: the error is not related to the kernel itself and is related to an attempt to autobuild on a recent change. The available builds are manually built/uploaded so the kernels should run without any issues

> Instead of manually managing complex dependencies, wrestling with compilation flags, or building libraries like Triton or CUTLASS from source, you can use the `kernels` library to instantly fetch and run pre-compiled, optimized kernels.

Author: agreed that is a strong argument, currently the flash_attn repo is actually building FA2 as it includes `mha_varlen_fwd` but not FA3 as the specialized H100 kernels are not built so I think I need to update the kernel repo to be more clear and support FA3. 

In the article i've added a small section that clearly compares using the kernel hub to fetch flash attention in seconds rather than building from source (hours) and link to the FA requirements (96GB RAM, may cores, etc). Please let me know if I should change it in any way!

## Get Started and Next Steps!

You've seen how easy it is to fetch and use optimized kernels with the Hugging Face Kernel Hub. Ready to try it yourself?

1.  **Install the library:**
    ```bash
    pip install kernels torch numpy
    ```
    Ensure you have a compatible PyTorch version and gpu driver installed.

2.  **Browse the Hub:** Explore available kernels on the Hugging Face Hub under the [`kernels` tag](https://huggingface.co/kernels) or within organizations like [`kernels-community`](https://huggingface.co/kernels-community). Look for kernels relevant to your operations (activations, attention, normalization like LayerNorm/RMSNorm, etc.).

3.  **Experiment:** Try replacing components in your own models. Use `get_kernel("user-or-org/kernel-name")`. **Crucially, inspect the loaded kernel object** (e.g., `print(dir(loaded_kernel))`) or check its Hub repository documentation to understand how to correctly call its functions/methods and what parameters (weights, biases, inputs, epsilon) it expects.

4.  **Benchmark:** Measure the performance impact on your specific hardware and workload. Don't forget to check for numerical correctness (`torch.testing.assert_close`).

> 5.  **(Advanced) Contribute:** If you develop optimized kernels, consider sharing them on the Hub!

Reviewer 5: Maybe here open an issue in the kernels repo/ kernels org on the hub so that people can request some kernels as well?

Reviewer 5: On second thought, issue/ discussion on kernels org on the hub would be even better.

Author: agreed I think it may be best to have those conversations in discussions

Reviewer 4: I'm curious how they're being shared as well, might be nice to include here without letting people dig somewhere else

Reviewer 5: created a discussion here: https://huggingface.co/spaces/kernels-community/README/discussions/1

---
title: "Learn the Hugging Face Kernel Hub in 5 Minutes"
thumbnail: /blog/assets/hello-hf-kernels/kernel-hub-five-mins-short.png
authors:
- user: drbh
- user: danieldk
- user: narsil
- user: pcuenca
- user: pagezyhf
- user: merve
- user: reach-vb
date: 2025-03-28 
---

# 🏎️ Enhance Your Models in 5 Minutes with the Hugging Face Kernel Hub

**Boost your model performance with pre-optimized kernels, easily loaded from the Hub.**

Today, we'll explore an exciting development from Hugging Face: the **Kernel Hub**! As ML practitioners, we know that maximizing performance often involves diving deep into optimized code, custom CUDA kernels, or complex build systems. The Kernel Hub simplifies this process dramatically!

Below is a short example of how to use a kernel in your code.

> ```python

Reviewer 5: suggestion to make snippets complete w/ pip install instructions (you can also link a colab notebook here too)

import torch

from kernels import get_kernel

# Download optimized kernels from the Hugging Face hub
activation = get_kernel("kernels-community/activation")

# Random tensor
x = torch.randn((10, 10), dtype=torch.float16, device="cuda")

# Run the kernel
y = torch.empty_like(x)
activation.gelu_fast(y, x)

print(y)
```

In the next sections we'll cover the following topics:

1.  **What is the Kernel Hub?** - Understanding the core concept.
2.  **How to use the Kernel Hub** - A quick code example.
3.  **Adding a Kernel to a Simple Model** - A practical integration using RMSNorm.
4.  **Reviewing Performance Impact** - Benchmarking the RMSNorm difference.
5.  **Real world use cases** - Examples of how the kernels library is being used in other projects.

We'll introduce these concepts quickly – the core idea can be grasped in about 5 minutes (though experimenting and benchmarking might take a bit longer!).

## 1. What is the Kernel Hub?

> The [Kernel Hub](https://huggingface.co/kernels-community) (👈 Check it out!) allows Python libraries and applications to **load optimized compute kernels directly from the Hugging Face Hub**. Think of it like the Model Hub, but for low-level, high-performance code snippets (kernels) that accelerate specific operations, often on GPUs. 

Reviewer 5: ```suggestion
The [Kernel Hub](https://huggingface.co/kernels-community) (👈 Follow it to keep up with latest kernels!) allows Python libraries and applications to **load optimized compute kernels directly from the Hugging Face Hub**. Think of it like the Model Hub, but for low-level, high-performance code snippets (kernels) that accelerate specific operations, often on GPUs. 
```
maybe
